File: hygmap-frontend/src/types/signal.ts
/**
 * Type definitions for SETI signal overlay
 */

export type SignalType = 'transmit' | 'receive'

export interface Signal {
  id: number
  name?: string | null
  type: SignalType
  time?: string | null
  ra?: number | null
  dec?: number | null
  frequency?: number | null
  notes?: string | null
  x: number
  y: number
  z: number
  last_updated?: string | null
  display_name: string
}

export interface SignalListResponse {
  result: string
  data: Signal[]
  length: number
}

export function getSignalDisplayName(signal: Signal): string {
  return signal.display_name || signal.name || `Signal ${signal.id}`
}

export function getSignalAgeYears(signal: Signal): number | null {
  if (!signal.time) return null
  const timestamp = new Date(signal.time).getTime()
  if (Number.isNaN(timestamp)) return null
  const diffMs = Date.now() - timestamp
  return diffMs / (1000 * 60 * 60 * 24 * 365.25)
}

File: hygmap-api/app/schemas/signal.py
"""
Pydantic schemas for SETI signal data
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field, computed_field

SignalType = Literal["transmit", "receive"]


class Signal(BaseModel):
    """Signal row returned from the signals table"""

    id: int
    name: Optional[str] = None
    type: SignalType
    time: Optional[datetime] = None
    ra: Optional[float] = None
    dec: Optional[float] = None
    frequency: Optional[float] = None
    notes: Optional[str] = None
    x: float = Field(..., description="Galactic X coordinate in parsecs")
    y: float = Field(..., description="Galactic Y coordinate in parsecs")
    z: float = Field(..., description="Galactic Z coordinate in parsecs")
    last_updated: Optional[datetime] = None

    @computed_field
    @property
    def display_name(self) -> str:
        """Fallback display name if the signal lacks a label"""
        return self.name or f"Signal {self.id}"


class SignalListResponse(BaseModel):
    """Response payload for signal list queries"""

    result: str = "success"
    data: list[Signal]
    length: int
